omisego/elixir-omg
# add timestamps to all table and fix tables that currently have timest…

## Overview

Add timestamps to all tables that support UTC times with time zone to microsecond granularity.

This is per general best practice. For more info please reference: https://wiki.postgresql.org/wiki/Don't_Do_This#Don.27t_use_timestamp_.28without_time_zone.29_to_store_UTC_times

## Changes

- Fix the way migrations are done. Ecto's timestamp stuff is poorly named and works in a very unexpected way.
  - Migration files need to use `timestamps([type: :timestamptz, default: "1970-01-01T00:00:00Z"])` to get the correct column type. Note the `:timestamptz`type
  - The actual db model files need to use: `timestamps(type: :utc_datetime_usec)`
- New migration file added to fix tables already containing timestamps and also to correctly add timestamps to existing tables.
- Timestamps removed from join table `ethevents_txoutputs` because ecto has poor support for join tables with extra columns.
- Update all db model files to have: `timestamps(type: :utc_datetime_usec)`

## Testing

1. Start with database containing `blocks` and `transactions` rows that already exist, but have no `inserted_at` and `updated_at` columns.
2. Also start with the database containing `ethevents` and `txoutputs` rows that already exist, and also have `inserted_at` and `updated_at` columns.
3. Run the migration
4. Inspect the `blocks` and `transactions` tables to confirm NON NULL constraints were not violated and they contain the current default values of `1970-01-01T00:00:00Z`.
5. Create new transactions and blocks and confirm they contain correct UTC `inserted_at` and `updated_at` data equaling 'now'.
6. Run a 'fake' test that updates the `ethevents` table to confirm that the `updated_at` column gets updated when the row is updated.
